/********************************************************************************
**
**	Copyright (c) 2007-2014 Witold Gantzke & Kirill Lepskiy
**
**	This file is part of the ACF-Solutions Toolkit.
**
**	This file may be used under the terms of the GNU Lesser
**	General Public License version 2.1 as published by the Free Software
**	Foundation and appearing in the file LicenseLGPL.txt included in the
**	packaging of this file.  Please review the following information to
**	ensure the GNU Lesser General Public License version 2.1 requirements
**	will be met: http://www.gnu.org/licenses/old-licenses/lgpl-2.1.html.
**
**	If you are unsure which license is appropriate for your use, please
**	contact us at info@imagingtools.de.
**
** 	See http://www.ilena.org, write info@imagingtools.de or contact
**	by Skype to ACF_infoline for further information about the ACF-Solutions.
**
********************************************************************************/


#include "ihotfgui/CDirectoryMonitorParamsGui.h"


// Qt includes
#if QT_VERSION >= 0x050000
#include <QtWidgets/QHeaderView>
#else
#include <QtGui/QHeaderView>
#endif

// ACF includes
#include "istd/TChangeNotifier.h"
#include "iqt/CSignalBlocker.h"
#include "iwidgets/CItemDelegate.h"


namespace ihotfgui
{


// reimplemented (imod::IModelEditor)

void CDirectoryMonitorParamsGui::UpdateModel() const
{
	Q_ASSERT(IsGuiCreated());

	ihotf::IDirectoryMonitorParams* objectPtr = GetObjectPtr();
	Q_ASSERT(objectPtr != NULL);

	objectPtr->SetPoolingIntervall(PoolingIntervallSpin->value());

	int observedItems = 0;
	if (FilesCheck->isChecked()){
		observedItems = observedItems | ihotf::IDirectoryMonitorParams::OI_FILES;
	}

	if (DirsCheck->isChecked()){
		observedItems = observedItems | ihotf::IDirectoryMonitorParams::OI_DIR;
	}

	if (DrivesCheck->isChecked()){
		observedItems = observedItems | ihotf::IDirectoryMonitorParams::OI_DRIVES;
	}

	objectPtr->SetObservedItemTypes(observedItems);

	int observedChanges = 0;
	if (ModifiedCheck->isChecked()){
		observedChanges = observedChanges | ihotf::IDirectoryMonitorParams::OC_MODIFIED;
	}

	if (AddedCheck->isChecked()){
		observedChanges = observedChanges | ihotf::IDirectoryMonitorParams::OC_ADD;
	}

	if (RemovedCheck->isChecked()){
		observedChanges = observedChanges | ihotf::IDirectoryMonitorParams::OC_REMOVE;
	}

	if (AttributesCheck->isChecked()){
		observedChanges = observedChanges | ihotf::IDirectoryMonitorParams::OC_ATTR_CHANGED;
	}

	objectPtr->SetObservedChanges(observedChanges);

	QTreeWidgetItemIterator acceptIter(AcceptPatternsList);
	QStringList acceptPatterns;
	while (*acceptIter){
		acceptPatterns.push_back((*acceptIter)->text(0));

		++acceptIter;
	}

	objectPtr->SetAcceptPatterns(acceptPatterns);

	QTreeWidgetItemIterator ignoreIter(IgnorePatternsList);
	QStringList ignorePatterns;
	while (*ignoreIter){
		ignorePatterns.push_back((*ignoreIter)->text(0));

		++ignoreIter;
	}

	objectPtr->SetIgnorePatterns(ignorePatterns);
}

	
// reimplemenented (iqtgui::TGuiObserverWrap)

void CDirectoryMonitorParamsGui::UpdateGui(int /*updateFlags*/)
{
	Q_ASSERT(IsGuiCreated());

	ihotf::IDirectoryMonitorParams* objectPtr = GetObjectPtr();
	if (objectPtr != NULL){
		ResetEditor();

		PoolingIntervallSpin->setValue(objectPtr->GetPoolingIntervall());

		int observedItems = objectPtr->GetObservedItemTypes();
		FilesCheck->setChecked((observedItems & ihotf::IDirectoryMonitorParams::OI_FILES) != 0);
		DirsCheck->setChecked((observedItems & ihotf::IDirectoryMonitorParams::OI_DIR) != 0);
		DrivesCheck->setChecked((observedItems & ihotf::IDirectoryMonitorParams::OI_DRIVES) != 0);

		int observedChanges = objectPtr->GetObservedChanges();
		ModifiedCheck->setChecked((observedChanges & ihotf::IDirectoryMonitorParams::OC_MODIFIED) != 0);
		AddedCheck->setChecked((observedChanges & ihotf::IDirectoryMonitorParams::OC_ADD) != 0);
		RemovedCheck->setChecked((observedChanges & ihotf::IDirectoryMonitorParams::OC_REMOVE) != 0);
		AttributesCheck->setChecked((observedChanges & ihotf::IDirectoryMonitorParams::OC_ATTR_CHANGED) != 0);

		QStringList acceptPatterns = objectPtr->GetAcceptPatterns();
		for (int patternIndex = 0; patternIndex < int(acceptPatterns.size()); patternIndex++){
			AddPattern(acceptPatterns[patternIndex], AcceptPatternsList);
		}

		QStringList ignorePatterns = objectPtr->GetIgnorePatterns();
		for (int patternIndex = 0; patternIndex < int(ignorePatterns.size()); patternIndex++){
			AddPattern(ignorePatterns[patternIndex], IgnorePatternsList);
		}
	}
}


// reimplemented (iqtgui::CGuiComponentBase)

void CDirectoryMonitorParamsGui::OnGuiCreated()
{
	BaseClass::OnGuiCreated();

	AcceptPatternsList->setEditTriggers(QAbstractItemView::DoubleClicked |  QAbstractItemView::EditKeyPressed);

	AcceptPatternsList->setItemDelegate(new iwidgets::CItemDelegate(20, AcceptPatternsList));

	IgnorePatternsList->setEditTriggers(QAbstractItemView::DoubleClicked |  QAbstractItemView::EditKeyPressed);

	IgnorePatternsList->setItemDelegate(new iwidgets::CItemDelegate(20, AcceptPatternsList));
}


// protected slots

void CDirectoryMonitorParamsGui::on_FilesCheck_toggled(bool/* isChecked*/)
{
	OnModelUpdate();
}


void CDirectoryMonitorParamsGui::on_DirsCheck_toggled(bool/* isChecked*/)
{
	OnModelUpdate();
}


void CDirectoryMonitorParamsGui::on_DrivesCheck_toggled(bool/* isChecked*/)
{
	OnModelUpdate();
}


void CDirectoryMonitorParamsGui::on_ModifiedCheck_toggled(bool/* isChecked*/)
{
	OnModelUpdate();
}


void CDirectoryMonitorParamsGui::on_AddedCheck_toggled(bool/* isChecked*/)
{
	OnModelUpdate();
}


void CDirectoryMonitorParamsGui::on_RemovedCheck_toggled(bool/* isChecked*/)
{
	OnModelUpdate();
}


void CDirectoryMonitorParamsGui::on_AttributesCheck_toggled(bool/* isChecked*/)
{
	OnModelUpdate();
}


void CDirectoryMonitorParamsGui::on_PoolingIntervallSpin_valueChanged(double/* isChecked*/)
{
	OnModelUpdate();
}


void CDirectoryMonitorParamsGui::on_AddAcceptPatternButton_clicked()
{
	AddPattern("<Filter>", AcceptPatternsList);

	OnModelUpdate();
}


void CDirectoryMonitorParamsGui::on_RemoveAcceptPatternButton_clicked()
{
	RemoveSelectedPatterns(AcceptPatternsList);
}


void CDirectoryMonitorParamsGui::on_AcceptPatternsList_itemSelectionChanged()
{
	QList<QTreeWidgetItem*> selectedItems = AcceptPatternsList->selectedItems();

	RemoveAcceptPatternButton->setEnabled(!selectedItems.isEmpty());
}


void CDirectoryMonitorParamsGui::on_AcceptPatternsList_itemChanged(QTreeWidgetItem* /*item*/, int /*column*/)
{
	OnModelUpdate();
}


void CDirectoryMonitorParamsGui::on_AddIgnorePatternButton_clicked()
{
	AddPattern("<Filter>", IgnorePatternsList);

	OnModelUpdate();
}


void CDirectoryMonitorParamsGui::on_RemoveIgnorePatternButton_clicked()
{
	RemoveSelectedPatterns(IgnorePatternsList);
}


void CDirectoryMonitorParamsGui::on_IgnorePatternsList_itemSelectionChanged()
{
	QList<QTreeWidgetItem*> selectedItems = IgnorePatternsList->selectedItems();

	RemoveIgnorePatternButton->setEnabled(!selectedItems.isEmpty());
}


void CDirectoryMonitorParamsGui::on_IgnorePatternsList_itemChanged(QTreeWidgetItem* /*item*/, int /*column*/)
{
	OnModelUpdate();
}


// private methods

void CDirectoryMonitorParamsGui::AddPattern(const QString& filter, QTreeWidget* treeView)
{
	iqt::CSignalBlocker block(treeView);

	QTreeWidgetItem* newFilterItem = new QTreeWidgetItem(treeView);
	newFilterItem->setText(0, filter);
	newFilterItem->setFlags(Qt::ItemIsEditable | Qt::ItemIsEnabled | Qt::ItemIsSelectable);

	treeView->addTopLevelItem(newFilterItem);
}


void CDirectoryMonitorParamsGui::RemoveSelectedPatterns(QTreeWidget* treeView)
{
	QList<QTreeWidgetItem*> selectedItems = treeView->selectedItems();

	if (!selectedItems.isEmpty()){
		for (QList<QTreeWidgetItem*>::iterator index = selectedItems.begin(); index != selectedItems.end(); index++){
			QTreeWidgetItem* itemPtr = treeView->takeTopLevelItem(AcceptPatternsList->indexOfTopLevelItem(*index));

			if (itemPtr != NULL){
				delete itemPtr;
			}
		}

		OnModelUpdate();
	}
}


void CDirectoryMonitorParamsGui::ResetEditor()
{
	while (AcceptPatternsList->topLevelItemCount() > 0){
		delete AcceptPatternsList->takeTopLevelItem(0);
	}	

	while (IgnorePatternsList->topLevelItemCount() > 0){
		delete IgnorePatternsList->takeTopLevelItem(0);
	}	
}


void CDirectoryMonitorParamsGui::OnModelUpdate()
{
	DoUpdateModel();
}


} // namespace ihotfgui


